### PR TITLE
Issue/2571 masterbar reader fixes

### PR DESCRIPTION
**Pull request by nbradbury**
This PR changes the logic for when the reader post list is automatically refreshed to show newly downloaded posts and how the "New Posts" message is handled. 

In the past, the list was always refreshed when there are new posts and the "new posts" message would appear to let the user know about them. This PR changes the logic so that the "new posts" message always appears, and the new posts aren't loaded until the user taps that message. This solves a number of small (but annoying) issues related to the list being refreshed while the user was reading it (including #2571).

### CODE REVIEW

**Comment by roundhill** [comment]
Hmm, I think I like how it worked before better. It was nice to just return to the app and then start scrolling up. Now it seems like I'll have to tap the button every time, which seems like a step backwards IMO.

Can you clarify what some of the annoying issues were? I never really noticed anything major.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java [hunk 1]
```diff
@@ -538,12 +540,13 @@ public void onScrollCompleted() {
...
-     * toolbar is enabled since it can intercept the touch event and pass it on to the swipe layout,
-     * cause a refresh to begin even though the user is simply scrolling up the posts
+     * disable swipe-to-refresh if "new posts" bar is showing or posts can be scrolled up
+     * because otherwise it can intercept the touch event and pass it on to the swipe layout,
+     * causing a refresh to begin even though the user is simply scrolling up the posts.
+     * note that this problem only applies when the fragment toolbar is enabled.
      */
     private void checkSwipeToRefresh() {
-        mSwipeToRefreshLayout.setEnabled(!mRecyclerView.canScrollUp());
+        mSwipeToRefreshLayout.setEnabled(!mRecyclerView.canScrollUp() && !isNewPostsBarShowing());
```
Comment (roundhill) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java]: I was still able to do a PTR after switching to the Freshly Pressed tag and the new posts bar appeared:

https://cloudup.com/iOx611BxnOc

Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java]: > I was still able to do a PTR after switching to the Freshly Pressed tag and the new posts bar appeared

Fixed in 21a8b60

**Comment by nbradbury** [comment]
> Can you clarify what some of the annoying issues were? I never really noticed anything major.

The biggest problem I was seeing was with "Blogs I Follow" when you have a mix of wp.com and RSS. The backend gets new wp.com posts immediately, but it _can_ be 24hrs or so before new RSS posts appear. Because posts are sorted by pubDate, this means the new RSS posts may appear further down the list rather than at the top. Not only can this result in you not seeing them, but it also can result in the list jumping as these posts are inserted (so you lose your place - most likely a bug in RecyclerView).

Sorting the posts by date received rather than date published is one way to prevent this, but it comes with its own set of issues ("why is the ordering different than the web?", etc.).

FWIW, once we switch to the latest support library I'm going to experiment with the new [SortedList](http://developer.android.com/reference/android/support/v7/util/SortedList.html) and see if it helps.

With all that said, I do agree it would be nice to skip that "New Posts" message and show the new posts right away. I'll see if I can tweak the logic of when that bar is shown to make it less annoying.

**Comment by nbradbury** [comment]
@roundhill In dc99507 I changed the logic so that the "new posts" message only appears if you've scrolled down the list, which is similar to how it used to work (plus it's simpler, which in my book is a vote in its favor). Do you think this change is better than always showing that message?

**Comment by roundhill** [comment]
:shipit:

**Pull request merged**